File: dart_sample/lib/demos/general/autofill_group_demo_page.dart
import 'package:flutter/services.dart';
import 'package:material_ui/material_ui.dart';

class AutofillGroupDemoPage extends StatefulWidget {
  const AutofillGroupDemoPage({super.key});

  @override
  State<AutofillGroupDemoPage> createState() => _AutofillGroupDemoPageState();
}

class _AutofillGroupDemoPageState extends State<AutofillGroupDemoPage> {
  late final TextEditingController _usernameController;
  late final TextEditingController _passwordController;
  late final TextEditingController _emailController;
  AutofillContextAction _onDisposeAction = AutofillContextAction.commit;
  bool _emailAutofillEnabled = true;
  String _lastAction = '(none)';

  @override
  void initState() {
    super.initState();
    _usernameController = TextEditingController();
    _passwordController = TextEditingController();
    _emailController = TextEditingController();
  }

  @override
  void dispose() {
    _usernameController.dispose();
    _passwordController.dispose();
    _emailController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'AutofillGroup',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Fields sharing the closest AutofillGroup are cross-referenced by the '
          'platform. Disposing the topmost group finishes the autofill context.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            SizedBox(
              width: 160,
              child: _MenuButton(
                label: _onDisposeAction == AutofillContextAction.commit
                    ? 'onDispose: commit'
                    : 'onDispose: cancel',
                onTap: () => setState(() {
                  _onDisposeAction =
                      _onDisposeAction == AutofillContextAction.commit
                      ? AutofillContextAction.cancel
                      : AutofillContextAction.commit;
                }),
                background: const Color(0xFFDCE3ED),
              ),
            ),
            SizedBox(
              width: 160,
              child: _MenuButton(
                label: _emailAutofillEnabled
                    ? 'Email: autofill on'
                    : 'Email: autofill off',
                onTap: () => setState(
                  () => _emailAutofillEnabled = !_emailAutofillEnabled,
                ),
                background: const Color(0xFFE9F5EC),
              ),
            ),
          ],
        ),
        SizedBox(
          width: 200,
          child: _MenuButton(
            label: 'finishAutofillContext',
            onTap: () {
              setState(() {
                TextInput.finishAutofillContext();
                _lastAction = 'finishAutofillContext(shouldSave: true)';
              });
            },
            background: const Color(0xFFF3E8D8),
          ),
        ),
        Text(
          'last action: $_lastAction',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        AutofillGroup(
          onDisposeAction: _onDisposeAction,
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            spacing: 8,
            children: <Widget>[
              const Text(
                'Username',
                style: TextStyle(fontSize: 12, color: Colors.black54),
              ),
              _buildTextField(
                controller: _usernameController,
                placeholder: 'username',
                autofillHints: const <String>[AutofillHints.username],
              ),
              const Text(
                'Password',
                style: TextStyle(fontSize: 12, color: Colors.black54),
              ),
              _buildTextField(
                controller: _passwordController,
                placeholder: 'password',
                obscureText: true,
                autofillHints: const <String>[AutofillHints.password],
              ),
              const Text(
                'Email',
                style: TextStyle(fontSize: 12, color: Colors.black54),
              ),
              _buildTextField(
                controller: _emailController,
                placeholder: 'email',
                autofillHints: _emailAutofillEnabled
                    ? const <String>[AutofillHints.email]
                    : null,
              ),
              Builder(
                builder: (BuildContext inner) => Text(
                  'clients in group: '
                  '${AutofillGroup.of(inner).autofillClients.length}',
                  style: const TextStyle(fontSize: 12, color: Colors.black),
                ),
              ),
            ],
          ),
        ),
        const Text(
          'Keyboard type and autocorrect are inferred from the first hint when '
          'they are not given explicitly.',
          style: TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
      ],
    );
  }

  Widget _buildTextField({
    required TextEditingController controller,
    required String placeholder,
    required List<String>? autofillHints,
    bool obscureText = false,
  }) {
    return TextField(
      controller: controller,
      obscureText: obscureText,
      autofillHints: autofillHints,
      decoration: InputDecoration(
        hintText: placeholder,
        isDense: true,
        filled: true,
        fillColor: const Color(0xFFE8F0FE),
        border: const OutlineInputBorder(),
      ),
    );
  }
}

class _MenuButton extends StatelessWidget {
  const _MenuButton({
    required this.label,
    required this.onTap,
    required this.background,
  });

  final String label;
  final VoidCallback onTap;
  final Color background;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        color: background,
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
        child: Text(
          label,
          style: const TextStyle(fontSize: 12, color: Colors.black),
        ),
      ),
    );
  }
}

File: src/Sample/Plumix.Sample/Demos/General/AutofillGroupDemoPage.cs
using System.Linq;
using Avalonia;
using Avalonia.Media;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/general/autofill_group_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class AutofillGroupDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new AutofillGroupDemoPageState();
    }
}

internal sealed class AutofillGroupDemoPageState : State
{
    private TextEditingController _usernameController = null!;
    private TextEditingController _passwordController = null!;
    private TextEditingController _emailController = null!;
    private AutofillContextAction _onDisposeAction = AutofillContextAction.Commit;
    private bool _emailAutofillEnabled = true;
    private string _lastAction = "(none)";

    public override void InitState()
    {
        _usernameController = new TextEditingController();
        _passwordController = new TextEditingController();
        _emailController = new TextEditingController();
    }

    public override void Dispose()
    {
        _usernameController.Dispose();
        _passwordController.Dispose();
        _emailController.Dispose();
    }

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("AutofillGroup", fontSize: 20, color: Colors.Black),
                new Text(
                    "Fields sharing the closest AutofillGroup are cross-referenced by the platform. "
                    + "Disposing the topmost group finishes the autofill context.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        new SizedBox(
                            width: 160,
                            child: new CounterTapButton(
                                label: _onDisposeAction == AutofillContextAction.Commit
                                    ? "onDispose: commit"
                                    : "onDispose: cancel",
                                onTap: () => SetState(() => _onDisposeAction =
                                    _onDisposeAction == AutofillContextAction.Commit
                                        ? AutofillContextAction.Cancel
                                        : AutofillContextAction.Commit),
                                background: Color.Parse("#FFDCE3ED"),
                                foreground: Colors.Black,
                                fontSize: 12,
                                padding: new Thickness(10, 8))),
                        new SizedBox(
                            width: 160,
                            child: new CounterTapButton(
                                label: _emailAutofillEnabled ? "Email: autofill on" : "Email: autofill off",
                                onTap: () => SetState(() =>
                                    _emailAutofillEnabled = !_emailAutofillEnabled),
                                background: Color.Parse("#FFE9F5EC"),
                                foreground: Colors.Black,
                                fontSize: 12,
                                padding: new Thickness(10, 8))),
                    ]),
                new SizedBox(
                    width: 200,
                    child: new CounterTapButton(
                        label: "finishAutofillContext",
                        onTap: () => SetState(() =>
                        {
                            TextInput.FinishAutofillContext();
                            _lastAction = "finishAutofillContext(shouldSave: true)";
                        }),
                        background: Color.Parse("#FFF3E8D8"),
                        foreground: Colors.Black,
                        fontSize: 12,
                        padding: new Thickness(10, 8))),
                new Text($"last action: {_lastAction}", fontSize: 12, color: Colors.DarkSlateGray),
                new AutofillGroup(
                    onDisposeAction: _onDisposeAction,
                    child: new Column(
                        crossAxisAlignment: CrossAxisAlignment.Stretch,
                        spacing: 8,
                        children:
                        [
                            new Text("Username", fontSize: 12, color: Colors.DimGray),
                            new EditableText(
                                controller: _usernameController,
                                placeholder: "username",
                                autofillHints: [AutofillHints.Username]),
                            new Text("Password", fontSize: 12, color: Colors.DimGray),
                            new EditableText(
                                controller: _passwordController,
                                placeholder: "password",
                                obscureText: true,
                                autofillHints: [AutofillHints.Password]),
                            new Text("Email", fontSize: 12, color: Colors.DimGray),
                            new EditableText(
                                controller: _emailController,
                                placeholder: "email",
                                autofillHints: _emailAutofillEnabled
                                    ? [AutofillHints.Email]
                                    : EditableText.AutofillDisabled),
                            new Builder(inner => new Text(
                                $"clients in group: {AutofillGroup.Of(inner).AutofillClients.Count()}",
                                fontSize: 12,
                                color: Colors.Black)),
                        ])),
                new Text(
                    "Keyboard type and autocorrect are inferred from the first hint when they are "
                    + "not given explicitly.",
                    fontSize: 12,
                    color: Colors.DarkSlateGray),
            ]);
    }
}
